### ISSUE TITLE
tensorflow mobile : pi_examples - unable to locate graph.pb.h

### ISSUE THREAD

**Issue by wfs** [tensorflow/tensorflow#3251]
GitHub issues are for bugs / installation problems / feature requests.  
For general support from the community, see [StackOverflow](https://stackoverflow.com/questions/tagged/tensorflow).
To make bugs and feature requests more easy to find and organize, we close issues that are deemed
out of scope for GitHub Issues and point people to StackOverflow.

For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.
### Environment info

Operating System:
pi@raspberrypi:~ $ cat /proc/version
Linux version 4.4.13-v7+ (dc4@dc4-XPS13-9333) (gcc version 4.9.3 (crosstool-NG crosstool-ng-1.22.0-88-g8460611) ) #894 SMP Mon Jun 13 13:13:27 BST 2016

Installed version of CUDA and cuDNN: 
(please attach the output of `ls -l /path/to/cuda/lib/libcud*`):

If installed from binary pip package, provide:
1. Which pip package you installed.
2. The output from `python -c "import tensorflow; print(tensorflow.__version__)"`.

If installed from sources, provide the commit hash:
### Steps to reproduce
1. Follow https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/pi_examples/
2. When you get to the final step in "Building the Examples", the following error occurs.
   pi@raspberrypi:~/tf/tensorflow $ make -f tensorflow/contrib/pi_examples/label_image/Makefile 
   gcc --std=c++11 -O0 -I/usr/local/include -I. -I/home/pi/tf/tensorflow/tensorflow/contrib/pi_examples/label_image/../../makefile/downloads -I/home/pi/tf/tensorflow/tensorflow/contrib/pi_examples/label_image/../../makefile/downloads/eigen-latest/ -I/home/pi/tf/tensorflow/tensorflow/contrib/pi_examples/label_image/../../makefile/gen/proto/ -I/home/pi/tf/tensorflow/tensorflow/contrib/pi_examples/label_image/../../makefile/gen/proto_text/ -c tensorflow/contrib/pi_examples/label_image/label_image.cc -o /home/pi/tf/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o
   tensorflow/contrib/pi_examples/label_image/label_image.cc:30:48: fatal error: tensorflow/core/framework/graph.pb.h: No such file or directory
   #include "tensorflow/core/framework/graph.pb.h"
                                                 ^
   compilation terminated.
   tensorflow/contrib/pi_examples/label_image/Makefile:78: recipe for target '/home/pi/tf/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o' failed
   make: **\* [/home/pi/tf/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o] Error 1
### What have you tried?
1. Searching for similar issues.
### Logs or other output that would be helpful

(If logs are large, please upload as attachment).

**Comment by jmchen-g** [comment]
@petewarden could you take a look at this please?

**Comment by petewarden** [comment]
Sorry for the long delay on this one. I've made some changes to the Pi building process and this seems to be working generally now. Can you let me know if you're still hitting problems?

**Comment by shaolinkhoa** [comment]
hi p @petewarden  ,
I have the same issue #4680 . when i fixed just like magicwifi said. I got the error
```
pi@raspberrypi:~/tensorflow/tensorflow $ make -f tensorflow/contrib/pi_examples/label_image/Makefile
gcc --std=c++11 -O0 -I/usr/local/include -I. -I/home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/../../makefile/downloads  -c tensorflow/contrib/pi_examples/label_image/label_image.cc -o /home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o
tensorflow/contrib/pi_examples/label_image/label_image.cc:31:48: fatal error: tensorflow/core/framework/graph.pb.h: No such file or directory
 #include "tensorflow/core/framework/graph.pb.h"
                                                ^
compilation terminated.
tensorflow/contrib/pi_examples/label_image/Makefile:80: recipe for target '/home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o' failed
make: *** [/home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o] Error 1
```
How i can fix this error for my pi zero

**Comment by drpngx** [comment]
Can you attach the generated makefile?

**Comment by shaolinkhoa** [comment]
Hi @drpngx 
[https://github.com/shaolinkhoa/tensorflow/blob/master/Makefile](url)

**Comment by drpngx** [comment]
400. Could you check the link?

**Comment by shaolinkhoa** [comment]
I fixed it

**Comment by drpngx** [comment]
Do you just want to install it?

There are some [instructions](https://github.com/samjabrahams/tensorflow-on-raspberry-pi) from one of our users @samjabrahams

**Comment by shaolinkhoa** [comment]
thank you for your information. i already tried,but their instruction didn't work cause it seems tensorflow only support armv7 pi2 and pi3, does not support armv6 like pi zero.

i try the instruction in pi_example and it worked but now i'm stuck on the missing graph like i said above

**Comment by drpngx** [comment]
It looks like there's a missing dependency, so I'd have to take a look at the graph. if you're building with `-j` just use one core first. Try `make -k` then `make` again

**Comment by shaolinkhoa** [comment]
hi, 
you mean 
`make -kf tensorflow/contrib/pi_examples/label_image/Makefile `        , right ?

it still the same error

**Comment by drpngx** [comment]
Can you remove the line with the comment? It's technically wrong, and IIRC in some versions of make the strict representation is honored.

**Comment by shaolinkhoa** [comment]
I tried 
`make -kf tensorflow/contrib/pi_examples/label_image/Makefile`
but the error is still the same

**```
pi@raspberrypi:~/tensorflow/tensorflow $ make -kf tensorflow/contrib/pi_examples/label_image/Makefile
gcc --std=c++11 -O0 -I/usr/local/include -I. -I/home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/../../makefile/downloads  -c tensorflow/contrib/pi_examples/label_image/label_image.cc -o /home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o
tensorflow/contrib/pi_examples/label_image/label_image.cc:31:48: fatal error: tensorflow/core/framework/graph.pb.h: No such file or directory
 #include "tensorflow/core/framework/graph.pb.h"
                                                ^
compilation terminated.
tensorflow/contrib/pi_examples/label_image/Makefile:80: recipe for target '/home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o' failed
make: *** [/home/pi/tensorflow/tensorflow/tensorflow/contrib/pi_examples/label_image/gen/obj/tensorflow/contrib/pi_examples/label_image/label_image.o] Error 1
make: Target 'all' not remade because of errors.
```**

**Comment by drpngx** [comment]
I don't see anything in that directory that creates the `graph.pb.h`, you have to call `protoc` yourself.

**Comment by shaolinkhoa** [comment]
I'm sorry. I really don't know what is `protoc` or how to create graph.pb.h ?

Would you mind giving more clues

**Comment by drpngx** [comment]
Look for protoc in `tensorflow/contrib/makefile/Makefile`

**Comment by shaolinkhoa** [comment]
Sir, please give me a solution to pass this problem so that I could install tensorflow on Pi zero.
i'm really not familiar with make.

I'm not the only one stucking with it. johnnymakk in post #4680 is having the same problem too

**Comment by drpngx** [comment]
Please copy the lines regarding protoc. `protoc` is the protobuf compiler that converts `.proto` files into `*.pb.{cc,h}`. It should be called to generate the `graph.pb.h`. The Makefile in `tensorflow/contrib/makefile/Makefile` contains this. It's not clear why it's missing from your contrib makefile, perhaps the aforementioned makefile was run first.

**Comment by johnnymakk** [comment]
Hi @drpngx thanks a lot for your help.
I think I'm following the same instructions as @shaolinkhoa .

Create a tensorflow static library using the makefile as in:
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/makefile#raspberry-pi
make -f tensorflow/contrib/makefile/Makefile HOST_OS=PI TARGET=PI \
 OPTFLAGS="-Os -mfpu=neon-vfpv4 -funsafe-math-optimizations -ftree-vectorize" CXX=g++-4.8

Yes, you are right, I ran the tensorflow/contrib/makefile/Makefile first to build tensorflow.
Seemed to build OK. No errors.

Then tried to compile the label_image example from makefile and received the graph.pb.h error.
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/pi_examples
make -f tensorflow/contrib/pi_examples/label_image/Makefile

Thanks a lot for any suggestions!

**Comment by drpngx** [comment]
@petewarden I'm not sure why the graph.pb.h is not generated.

**Comment by johnnymakk** [comment]
I managed to get the label image example compiled.
I cloned the most recent tensorflow repo and was able to make the project OK.
graph.pb.h is generated in tensorflow/contrib/makefile/gen/proto/tensorflow/core/framework
I noticed the fix for eigen is done.
Thanks @drpngx and @petewarden

**Comment by drpngx** [comment]
Could you summarize the steps that you used, so that other people can understand, in a nutshell, what the resolution is?

Thanks!

**Comment by johnnymakk** [comment]
I pretty much followed the instructions in the following posts.
First git clone the tensorflow repo. (did a fresh clone with the latest code)


https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/makefile#raspberry-pi
I used the second option which should optimize for Pi 2&3.
```
make -f tensorflow/contrib/makefile/Makefile HOST_OS=PI TARGET=PI \
OPTFLAGS="-Os -mfpu=neon-vfpv4 -funsafe-math-optimizations -ftree-vectorize" CXX=g++-4.8
```

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/pi_examples
`make -f tensorflow/contrib/pi_examples/label_image/Makefile`

I'm not sure what caused the problem with graph.pb.h but downloading the latest repo and going through the procedure again resolved the problem.

Thanks!

**Comment by drpngx** [comment]
Thanks!

**Comment by shaolinkhoa** [comment]
It seems Pi zero is hopeless. 
I followed all the steps again and it works just like johnnymakk said.

but when I run the line
`make -f tensorflow/contrib/pi_examples/label_image/Makefile`

I got error 
```
pi@raspberrypi:~/tensorflow $ tensorflow/contrib/pi_examples/label_image/gen/bin/label_image
Illegal instruction
```
Tensorflow seems cannot install on ARM V6
